File: Advanced DSA/02. Array Techniques/02. Equilibrium index of an array/Solution.py
class Solution:
    # @param A : list of integers
    # @return an integer
    def solve(self, A):
        suffix_sum = [0 for _ in range(len(A) + 1)]
        for i in range(len(A)-1, -1, -1):
            suffix_sum[i] = suffix_sum[i + 1] + A[i]
        
        prefix_sum = 0
        for i in range(len(A)):
            if prefix_sum == suffix_sum[i+1]:
                return i
            prefix_sum += A[i]
        return -1

File: Advanced DSA/02. Array Techniques/02. Equilibrium index of an array/Solution.cpp
int Solution::solve(vector<int> &A) {
    vector<int> suffix_sum(A.size() + 1, 0);
    for (int i = A.size() - 1; i >= 0; --i)
        suffix_sum[i] = suffix_sum[i+1] + A[i];
    
    int prefix_sum = 0;
    for (int i = 0; i < A.size(); ++i) {
        if (prefix_sum == suffix_sum[i + 1])
            return i;
        prefix_sum += A[i];
    }
    return -1;
}
